File: android-app/app/src/main/java/com/anonymous/chat/utils/PreferenceManager.java
package com.anonymous.chat.utils;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.Random;

public class PreferenceManager {
    private static final String PREF_NAME = "anonymous_chat_prefs";
    private static final String KEY_SERVER_HOST = "server_host";
    private static final String KEY_SERVER_PORT = "server_port";
    private static final String KEY_AUTH_KEY = "auth_key";
    private static final String KEY_RECOVERY_KEY = "recovery_key";
    private static final String KEY_TIMEZONE = "pref_timezone"; // "vn" or "utc"
    private static final String KEY_SOUND_ENABLED = "pref_sound_enabled";
    private static final String KEY_DEVICE_MAC = "chat_device_mac";

    public static final String DEFAULT_SERVER_HOST = "snow.pikamc.vn";
    public static final int DEFAULT_SERVER_PORT = 25222;

    private static PreferenceManager instance;
    private final SharedPreferences prefs;

    private PreferenceManager(Context context) {
        prefs = context.getApplicationContext().getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }

    public static synchronized PreferenceManager getInstance(Context context) {
        if (instance == null) {
            instance = new PreferenceManager(context);
        }
        return instance;
    }

    public String getDeviceMac() {
        String mac = prefs.getString(KEY_DEVICE_MAC, null);
        if (mac == null || mac.isEmpty()) {
            Random rand = new Random();
            StringBuilder sb = new StringBuilder("MAC");
            for (int i = 0; i < 6; i++) {
                sb.append(i == 0 ? "-" : ":");
                sb.append(String.format("%02X", rand.nextInt(256)));
            }
            mac = sb.toString();
            prefs.edit().putString(KEY_DEVICE_MAC, mac).apply();
        }
        return mac;
    }

    public String getServerHost() {
        return prefs.getString(KEY_SERVER_HOST, DEFAULT_SERVER_HOST);
    }

    public void setServerHost(String host) {
        prefs.edit().putString(KEY_SERVER_HOST, host).apply();
    }

    public int getServerPort() {
        return prefs.getInt(KEY_SERVER_PORT, DEFAULT_SERVER_PORT);
    }

    public void setServerPort(int port) {
        prefs.edit().putInt(KEY_SERVER_PORT, port).apply();
    }

    public String getAuthKey() {
        return prefs.getString(KEY_AUTH_KEY, null);
    }

    public void setAuthKey(String authKey) {
        prefs.edit().putString(KEY_AUTH_KEY, authKey).apply();
    }

    public String getRecoveryKey() {
        return prefs.getString(KEY_RECOVERY_KEY, null);
    }

    public void setRecoveryKey(String recoveryKey) {
        prefs.edit().putString(KEY_RECOVERY_KEY, recoveryKey).apply();
    }

    public String getServerBaseUrl() {
        String host = getServerHost().trim();
        int port = getServerPort();
        if (host.isEmpty()) host = DEFAULT_SERVER_HOST;
        if (port <= 0) port = DEFAULT_SERVER_PORT;

        if (!host.startsWith("http://") && !host.startsWith("https://")) {
            host = "http://" + host;
        }

        // If port is already present in host string (e.g. http://domain:25222), don't append
        String afterScheme = host.substring(host.indexOf("://") + 3);
        if (!afterScheme.contains(":")) {
            return host + ":" + port;
        }
        return host;
    }

    public String getTimezone() {
        return prefs.getString(KEY_TIMEZONE, "vn");
    }

    public void setTimezone(String tz) {
        prefs.edit().putString(KEY_TIMEZONE, tz).apply();
    }

    public boolean isSoundEnabled() {
        return prefs.getBoolean(KEY_SOUND_ENABLED, true);
    }

    public void setSoundEnabled(boolean enabled) {
        prefs.edit().putBoolean(KEY_SOUND_ENABLED, enabled).apply();
    }
}

File: ios-app/AnonymousChat/Services/PreferenceManager.swift
import Foundation

public class PreferenceManager: ObservableObject {
    public static let shared = PreferenceManager()

    private let defaults = UserDefaults.standard

    private let keyServerHost = "server_host"
    private let keyServerPort = "server_port"
    private let keyAuthKey = "auth_key"
    private let keyRecoveryKey = "recovery_key"
    private let keyTimezone = "pref_timezone"
    private let keyDeviceMac = "chat_device_mac"
    private let keySoundEnabled = "pref_sound_enabled"

    public static let defaultServerHost = "snow.pikamc.vn"
    public static let defaultServerPort = 25222

    @Published public var serverHost: String {
        didSet { defaults.set(serverHost, forKey: keyServerHost) }
    }

    @Published public var serverPort: Int {
        didSet { defaults.set(serverPort, forKey: keyServerPort) }
    }

    @Published public var authKey: String? {
        didSet { defaults.set(authKey, forKey: keyAuthKey) }
    }

    @Published public var recoveryKey: String? {
        didSet { defaults.set(recoveryKey, forKey: keyRecoveryKey) }
    }

    @Published public var timezone: String {
        didSet { defaults.set(timezone, forKey: keyTimezone) }
    }

    @Published public var isSoundEnabled: Bool {
        didSet { defaults.set(isSoundEnabled, forKey: keySoundEnabled) }
    }

    @Published public var deviceMac: String

    private init() {
        let savedHost = defaults.string(forKey: keyServerHost) ?? PreferenceManager.defaultServerHost
        let savedPort = defaults.integer(forKey: keyServerPort)
        self.serverHost = savedHost.isEmpty ? PreferenceManager.defaultServerHost : savedHost
        self.serverPort = savedPort > 0 ? savedPort : PreferenceManager.defaultServerPort
        self.authKey = defaults.string(forKey: keyAuthKey)
        self.recoveryKey = defaults.string(forKey: keyRecoveryKey)
        self.timezone = defaults.string(forKey: keyTimezone) ?? "vn"
        self.isSoundEnabled = defaults.object(forKey: keySoundEnabled) == nil ? true : defaults.bool(forKey: keySoundEnabled)

        // Initialize MAC address
        if let existingMac = defaults.string(forKey: keyDeviceMac), !existingMac.isEmpty {
            self.deviceMac = existingMac
        } else {
            var bytes = [String]()
            for _ in 0..<6 {
                let val = Int.random(in: 0...255)
                bytes.append(String(format: "%02X", val))
            }
            let newMac = "MAC-" + bytes.joined(separator: ":")
            defaults.set(newMac, forKey: keyDeviceMac)
            self.deviceMac = newMac
        }
    }

    public func getDeviceMac() -> String {
        if !deviceMac.isEmpty {
            return deviceMac
        }
        if let mac = defaults.string(forKey: keyDeviceMac), !mac.isEmpty {
            self.deviceMac = mac
            return mac
        }
        var bytes = [String]()
        for _ in 0..<6 {
            let val = Int.random(in: 0...255)
            bytes.append(String(format: "%02X", val))
        }
        let newMac = "MAC-" + bytes.joined(separator: ":")
        defaults.set(newMac, forKey: keyDeviceMac)
        self.deviceMac = newMac
        return newMac
    }

    public var serverBaseUrl: String {
        var host = serverHost.trimmingCharacters(in: .whitespacesAndNewlines)
        if host.isEmpty { host = PreferenceManager.defaultServerHost }
        let port = serverPort > 0 ? serverPort : PreferenceManager.defaultServerPort

        if !host.hasPrefix("http://") && !host.hasPrefix("https://") {
            host = "http://" + host
        }

        if let range = host.range(of: "://") {
            let afterScheme = String(host[range.upperBound...])
            if !afterScheme.contains(":") {
                return "\(host):\(port)"
            }
        }
        return host
    }
}
